File: packages/effect-math/scripts/fixtures/complex.py
"""Complex-number domain fixture generation from NumPy reference implementations."""

from __future__ import annotations

import math
from typing import Any

import numpy as np

from ._common import metadata


def generate(generated_at: str) -> list[dict[str, Any]]:
    cases: list[dict[str, Any]] = []

    # ── Arithmetic ──────────────────────────────────────────────────────
    _arith_pairs: list[tuple[str, complex, complex]] = [
        ("unit", 1 + 2j, 3 + 4j),
        ("pure-real", 5 + 0j, -3 + 0j),
        ("pure-imag", 0 + 2j, 0 - 7j),
        ("large", 1e8 + 1e8j, 2e8 - 3e8j),
        ("small", 1e-12 + 1e-12j, 2e-12 + 3e-12j),
    ]

    for label, a, b in _arith_pairs:
        cases.append(_binary_case(f"add-{label}", "add", a, b, a + b))
        cases.append(_binary_case(f"subtract-{label}", "subtract", a, b, a - b))
        cases.append(_binary_case(f"multiply-{label}", "multiply", a, b, a * b))
        cases.append(_binary_case(f"divide-{label}", "divide", a, b, a / b))

    # ── Unary ───────────────────────────────────────────────────────────
    _unary_inputs: list[tuple[str, complex]] = [
        ("unit", 3 + 4j),
        ("pure-real", -5 + 0j),
        ("pure-imag", 0 + 7j),
        ("negative-both", -2 - 3j),
    ]

    for label, z in _unary_inputs:
        cases.append(_unary_complex_case(f"conjugate-{label}", "conjugate", z, z.conjugate()))
        cases.append(_unary_scalar_case(f"abs-{label}", "abs", z, abs(z)))
        cases.append(_unary_scalar_case(f"arg-{label}", "arg", z, np.angle(z)))

    # ── Transcendental ──────────────────────────────────────────────────
    _trans_inputs: list[tuple[str, complex]] = [
        ("unit", 1 + 1j),
        ("pure-real-neg", -1.0 + 0j),
        ("pure-imag", 0 + 3j),
        ("general", 0.5 - 1.5j),
    ]

    for label, z in _trans_inputs:
        cases.append(_unary_complex_case(f"exp-{label}", "exp", z, np.exp(z)))
        cases.append(_unary_complex_case(f"log-{label}", "log", z, np.log(z)))
        cases.append(_unary_complex_case(f"sqrt-{label}", "sqrt", z, np.sqrt(z)))

    # pow with complex exponent
    _pow_pairs: list[tuple[str, complex, complex]] = [
        ("int-exp", 2 + 1j, 3 + 0j),
        ("complex-exp", 1 + 1j, 1 + 1j),
        ("neg-base", -1 + 0j, 0.5 + 0j),
    ]

    for label, base, exp in _pow_pairs:
        cases.append(_pow_case(f"pow-{label}", base, exp, base**exp))

    # ── Trigonometric ───────────────────────────────────────────────────
    _trig_inputs: list[tuple[str, complex]] = [
        ("unit", 1 + 1j),
        ("pure-imag", 0 + 2j),
        ("general", 0.5 - 0.5j),
    ]

    for label, z in _trig_inputs:
        cases.append(_unary_complex_case(f"sin-{label}", "sin", z, np.sin(z)))
        cases.append(_unary_complex_case(f"cos-{label}", "cos", z, np.cos(z)))
        cases.append(_unary_complex_case(f"tan-{label}", "tan", z, np.tan(z)))

    # ── Hyperbolic ──────────────────────────────────────────────────────
    _hyp_inputs: list[tuple[str, complex]] = [
        ("unit", 1 + 1j),
        ("pure-imag", 0 + 2j),
        ("general", -1 + 0.5j),
    ]

    for label, z in _hyp_inputs:
        cases.append(_unary_complex_case(f"sinh-{label}", "sinh", z, np.sinh(z)))
        cases.append(_unary_complex_case(f"cosh-{label}", "cosh", z, np.cosh(z)))
        cases.append(_unary_complex_case(f"tanh-{label}", "tanh", z, np.tanh(z)))

    # ── Polar ───────────────────────────────────────────────────────────
    _polar_inputs: list[tuple[str, complex]] = [
        ("unit", 1 + 1j),
        ("pure-real-neg", -3 + 0j),
        ("general", 3 - 4j),
    ]

    for label, z in _polar_inputs:
        r = abs(z)
        theta = float(np.angle(z))
        cases.append({
            "id": f"toPolar-{label}",
            "operation": "toPolar",
            "input": {"re": z.real, "im": z.imag},
            "expected": {"r": float(r), "theta": theta},
        })

    # ── Complex-step derivative ─────────────────────────────────────────
    h = 1e-20
    _deriv_fns: list[tuple[str, str]] = [
        ("square", "square"),
        ("cube", "cube"),
        ("sin", "sin"),
        ("exp", "exp"),
    ]
    _deriv_points: list[tuple[str, float]] = [
        ("at-0", 0.0),
        ("at-1", 1.0),
        ("at-2", 2.0),
    ]

    for fn_label, fn_name in _deriv_fns:
        for pt_label, x in _deriv_points:
            z = complex(x, h)
            if fn_name == "square":
                fval = z ** 2
            elif fn_name == "cube":
                fval = z ** 3
            elif fn_name == "sin":
                fval = np.sin(z)
            elif fn_name == "exp":
                fval = np.exp(z)
            else:
                raise ValueError(f"Unknown function: {fn_name}")

            cases.append({
                "id": f"complexDerivative-{fn_label}-{pt_label}",
                "operation": "complexDerivative",
                "input": {"fn": fn_name, "x": x},
                "expected": float(fval.imag / h),
            })

    return [
        {
            "fixture": "complex.arithmetic-parity",
            "file": "complex/arithmetic-parity.json",
            "metadata": metadata(generated_at),
            "payload": {
                "cases": cases,
            },
        }
    ]


# ── Helper factories ────────────────────────────────────────────────────


def _binary_case(
    case_id: str, operation: str, a: complex, b: complex, result: complex
) -> dict[str, Any]:
    return {
        "id": case_id,
        "operation": operation,
        "input": {
            "aRe": float(a.real),
            "aIm": float(a.imag),
            "bRe": float(b.real),
            "bIm": float(b.imag),
        },
        "expected": {"re": float(result.real), "im": float(result.imag)},
    }


def _unary_complex_case(
    case_id: str, operation: str, z: complex, result: complex
) -> dict[str, Any]:
    return {
        "id": case_id,
        "operation": operation,
        "input": {"re": float(z.real), "im": float(z.imag)},
        "expected": {"re": float(result.real), "im": float(result.imag)},
    }


def _unary_scalar_case(
    case_id: str, operation: str, z: complex, result: float
) -> dict[str, Any]:
    return {
        "id": case_id,
        "operation": operation,
        "input": {"re": float(z.real), "im": float(z.imag)},
        "expected": float(result),
    }


def _pow_case(
    case_id: str, base: complex, exp: complex, result: complex
) -> dict[str, Any]:
    return {
        "id": case_id,
        "operation": "pow",
        "input": {
            "baseRe": float(base.real),
            "baseIm": float(base.imag),
            "expRe": float(exp.real),
            "expIm": float(exp.imag),
        },
        "expected": {"re": float(result.real), "im": float(result.imag)},
    }

File: packages/effect-math/src/Complex.ts
/**
 * Complex arithmetic, polar conversion, trigonometry, and dense-vector operations.
 *
 * @since 0.1.0
 * @module
 */
import { Array, Chunk, Effect, Number, Schema, Tuple } from "effect"
import { dual } from "effect/Function"

import * as Arithmetic from "./internal/complex/arithmetic.js"
import * as Trigonometric from "./internal/complex/trigonometric.js"
import * as PolicyGuard from "./internal/policyGuard.js"
import * as Numeric from "./Numeric.js"

const encodeNumber = Schema.encodeSync(Schema.NumberFromString)

/**
 * A schema-backed Cartesian complex value. Components preserve the full IEEE
 * 754 number domain; validated operations use the finite refinements below.
 *
 * @since 0.1.0
 * @category models
 */
export class Complex extends Schema.Class<Complex>("Complex")(
  {
    /** Real component. */
    re: Schema.Number,
    /** Imaginary component. */
    im: Schema.Number
  },
  { identifier: "@scenesystems/effect-math/Complex/Complex" }
) {}

/**
 * Accepts a non-negative radius and an angle in radians.
 *
 * @since 0.4.0
 * @category schemas
 */
export const Polar = Schema.Tuple(
  Schema.Number.pipe(Schema.nonNegative()),
  Schema.Number
).annotations({ identifier: "@scenesystems/effect-math/Complex/Polar" })

/**
 * A polar `[radius, angle]` pair.
 *
 * @since 0.4.0
 * @category models
 */
export type Polar = typeof Polar.Type

/**
 * Accepts finite components derived from the canonical complex value fields.
 *
 * @since 0.1.0
 * @category schemas
 */
export const Input = Schema.Struct({
  re: Complex.fields.re.pipe(Schema.finite()),
  im: Complex.fields.im.pipe(Schema.finite())
}).annotations({ identifier: "@scenesystems/effect-math/Complex/Input" })

/**
 * Finite Cartesian input for validated unary operations.
 *
 * @since 0.1.0
 * @category models
 */
export type Input = typeof Input.Type

/**
 * Accepts two finite Cartesian operands using the canonical component fields.
 *
 * @since 0.1.0
 * @category schemas
 */
export const BinaryInput = Schema.Struct({
  aRe: Input.fields.re,
  aIm: Input.fields.im,
  bRe: Input.fields.re,
  bIm: Input.fields.im
}).annotations({ identifier: "@scenesystems/effect-math/Complex/BinaryInput" })

/**
 * A finite pair of Cartesian operands for validated binary operations.
 *
 * @since 0.1.0
 * @category models
 */
export type BinaryInput = typeof BinaryInput.Type

/**
 * Reports malformed input to a validated complex operation.
 *
 * @since 0.1.0
 * @category errors
 */
export class DecodeError extends Schema.TaggedError<DecodeError>()("ComplexDecodeError", {
  /** Operation whose input failed decoding. */
  operation: Schema.String,
  /** Effect Schema issue report. */
  message: Schema.String
}) {}

/**
 * Reports a non-finite result rejected by strict precision.
 *
 * @since 0.1.0
 * @category errors
 */
export class DomainViolationError extends Schema.TaggedError<DomainViolationError>()("ComplexDomainViolationError", {
  /** Policy-aware operation that produced the result. */
  operation: Schema.String,
  /** Diagnostic describing the rejected result. */
  message: Schema.String
}) {}

/**
 * Failures emitted by validated and policy-aware complex operations.
 *
 * @since 0.1.0
 * @category errors
 */
export type OperationError = DecodeError | DomainViolationError

const decode = <A, I, R>(schema: Schema.Schema<A, I, R>, operation: string, input: unknown) =>
  Schema.decodeUnknown(schema)(input, { onExcessProperty: "error" }).pipe(
    Effect.mapError((error) => new DecodeError({ operation, message: error.message }))
  )

const fromCartesian = (components: Arithmetic.Cartesian): Complex =>
  new Complex({ re: Tuple.getFirst(components), im: Tuple.getSecond(components) })

/**
 * Constructs a complex value without normalization or finiteness checks.
 *
 * @since 0.1.0
 * @category constructors
 */
export const make = (re: number, im: number): Complex => new Complex({ re, im })

/**
 * Constructs a value on the real axis.
 *
 * @since 0.1.0
 * @category constructors
 */
export const fromReal = (re: number): Complex => new Complex({ re, im: 0 })

/**
 * Constructs a value on the imaginary axis.
 *
 * @since 0.1.0
 * @category constructors
 */
export const fromImaginary = (im: number): Complex => new Complex({ re: 0, im })

/**
 * Constructs a value from a non-negative radius and angle in radians. The
 * trusted operation does not validate the radius at runtime.
 *
 * @since 0.1.0
 * @category constructors
 */
export const fromPolar = (radius: number, angle: number): Complex =>
  new Complex({
    re: Number.multiply(radius, Numeric.cos(angle)),
    im: Number.multiply(radius, Numeric.sin(angle))
  })

/**
 * The additive identity.
 *
 * @since 0.1.0
 * @category constants
 */
export const zero = new Complex({ re: 0, im: 0 })

/**
 * The multiplicative identity.
 *
 * @since 0.1.0
 * @category constants
 */
export const one = new Complex({ re: 1, im: 0 })

/**
 * The imaginary unit.
 *
 * @since 0.1.0
 * @category constants
 */
export const i = new Complex({ re: 0, im: 1 })

/**
 * Adds `that` to `self`, in data-first or data-last form.
 *
 * @since 0.1.0
 * @category operations
 */
export const add: {
  (that: Complex): (self: Complex) => Complex
  (self: Complex, that: Complex): Complex
} = dual(
  2,
  (self: Complex, that: Complex): Complex => fromCartesian(Arithmetic.add(self.re, self.im, that.re, that.im))
)

/**
 * Subtracts `that` from `self`, in data-first or data-last form.
 *
 * @since 0.1.0
 * @category operations
 */
export const subtract: {
  (that: Complex): (self: Complex) => Complex
  (self: Complex, that: Complex): Complex
} = dual(
  2,
  (self: Complex, that: Complex): Complex => fromCartesian(Arithmetic.subtract(self.re, self.im, that.re, that.im))
)

/**
 * Multiplies `self` by `that`, in data-first or data-last form.
 *
 * @since 0.1.0
 * @category operations
 */
export const multiply: {
  (that: Complex): (self: Complex) => Complex
  (self: Complex, that: Complex): Complex
} = dual(
  2,
  (self: Complex, that: Complex): Complex => fromCartesian(Arithmetic.multiply(self.re, self.im, that.re, that.im))
)

/**
 * Divides `self` by `that` with Smith's overflow-resistant method. A zero
 * divisor produces `NaN` components rather than a typed failure.
 *
 * @since 0.1.0
 * @category operations
 */
export const divide: {
  (that: Complex): (self: Complex) => Complex
  (self: Complex, that: Complex): Complex
} = dual(
  2,
  (self: Complex, that: Complex): Complex => fromCartesian(Arithmetic.divide(self.re, self.im, that.re, that.im))
)

/**
 * Reflects a value across the real axis.
 *
 * @since 0.1.0
 * @category operations
 */
export const conjugate = (value: Complex): Complex => fromCartesian(Arithmetic.conjugate(value.re, value.im))

/**
 * Computes the modulus with an overflow-resistant hypotenuse operation.
 *
 * @since 0.1.0
 * @category operations
 */
export const abs = (value: Complex): number => Arithmetic.abs(value.re, value.im)

/**
 * Computes the principal phase angle in `(-π, π]`.
 *
 * @since 0.1.0
 * @category operations
 */
export const arg = (value: Complex): number => Arithmetic.arg(value.re, value.im)

/**
 * Computes `exp(a + bi) = exp(a)(cos(b) + i sin(b))`.
 *
 * @since 0.1.0
 * @category operations
 */
export const exp = (value: Complex): Complex => fromCartesian(Arithmetic.exp(value.re, value.im))

/**
 * Computes the principal complex logarithm with argument in `(-π, π]`.
 * Zero maps to `-Infinity + 0i`.
 *
 * @since 0.1.0
 * @category operations
 */
export const log = (value: Complex): Complex => fromCartesian(Arithmetic.log(value.re, value.im))

/**
 * Raises `self` to a complex exponent in data-first or data-last form. The
 * operation defines `0^0` as one and `0^w` as zero for nonzero `w`.
 *
 * @since 0.1.0
 * @category operations
 */
export const pow: {
  (exponent: Complex): (self: Complex) => Complex
  (self: Complex, exponent: Complex): Complex
} = dual(
  2,
  (self: Complex, exponent: Complex): Complex =>
    fromCartesian(Arithmetic.pow(self.re, self.im, exponent.re, exponent.im))
)

/**
 * Computes the principal complex square root. Negative real values produce a
 * non-negative imaginary component.
 *
 * @since 0.1.0
 * @category operations
 */
export const sqrt = (value: Complex): Complex => fromCartesian(Arithmetic.sqrt(value.re, value.im))

/**
 * Converts a Cartesian value to `[radius, principalAngle]`.
 *
 * @since 0.1.0
 * @category operations
 */
export const toPolar = (value: Complex): Polar =>
  Tuple.make(Arithmetic.abs(value.re, value.im), Arithmetic.arg(value.re, value.im))

/**
 * Computes complex sine from the analytic continuation of the real function.
 *
 * @since 0.1.0
 * @category operations
 */
export const sin = (value: Complex): Complex => fromCartesian(Trigonometric.sin(value.re, value.im))

/**
 * Computes complex cosine from the analytic continuation of the real function.
 *
 * @since 0.1.0
 * @category operations
 */
export const cos = (value: Complex): Complex => fromCartesian(Trigonometric.cos(value.re, value.im))

/**
 * Computes `sin(z) / cos(z)`; poles produce `NaN` components.
 *
 * @since 0.1.0
 * @category operations
 */
export const tan = (value: Complex): Complex => fromCartesian(Trigonometric.tan(value.re, value.im))

/**
 * Computes complex hyperbolic sine.
 *
 * @since 0.1.0
 * @category operations
 */
export const sinh = (value: Complex): Complex => fromCartesian(Trigonometric.sinh(value.re, value.im))

/**
 * Computes complex hyperbolic cosine.
 *
 * @since 0.1.0
 * @category operations
 */
export const cosh = (value: Complex): Complex => fromCartesian(Trigonometric.cosh(value.re, value.im))

/**
 * Computes `sinh(z) / cosh(z)`; poles produce `NaN` components.
 *
 * @since 0.1.0
 * @category operations
 */
export const tanh = (value: Complex): Complex => fromCartesian(Trigonometric.tanh(value.re, value.im))

/**
 * Computes the sesquilinear inner product `Σ conj(selfᵢ) * thatᵢ`. Inputs are
 * zipped, so unmatched trailing elements are ignored.
 *
 * @since 0.1.0
 * @category operations
 */
export const dot: {
  (that: Chunk.Chunk<Complex>): (self: Chunk.Chunk<Complex>) => Complex
  (self: Chunk.Chunk<Complex>, that: Chunk.Chunk<Complex>): Complex
} = dual(2, (self: Chunk.Chunk<Complex>, that: Chunk.Chunk<Complex>): Complex => {
  const initial: Arithmetic.Cartesian = Tuple.make(0, 0)
  const result = Chunk.zipWith(self, that, (left, right) => {
    const conjugated = Arithmetic.conjugate(left.re, left.im)
    return Arithmetic.multiply(Tuple.getFirst(conjugated), Tuple.getSecond(conjugated), right.re, right.im)
  }).pipe(
    Chunk.reduce(initial, (accumulator, value) =>
      Arithmetic.add(
        Tuple.getFirst(accumulator),
        Tuple.getSecond(accumulator),
        Tuple.getFirst(value),
        Tuple.getSecond(value)
      ))
  )
  return fromCartesian(result)
})

/**
 * Computes `sqrt(Σ |zᵢ|²)` for a complex vector.
 *
 * @since 0.1.0
 * @category operations
 */
export const norm = (values: Chunk.Chunk<Complex>): number =>
  Numeric.sqrt(
    Chunk.reduce(values, 0, (accumulator, value) =>
      Number.sum(
        accumulator,
        Number.sum(Number.multiply(value.re, value.re), Number.multiply(value.im, value.im))
      ))
  )

/**
 * Multiplies every vector element by a complex scalar.
 *
 * @since 0.1.0
 * @category operations
 */
export const scale: {
  (scalar: Complex): (self: Chunk.Chunk<Complex>) => Chunk.Chunk<Complex>
  (self: Chunk.Chunk<Complex>, scalar: Complex): Chunk.Chunk<Complex>
} = dual(
  2,
  (self: Chunk.Chunk<Complex>, scalar: Complex): Chunk.Chunk<Complex> =>
    Chunk.map(self, (value) => fromCartesian(Arithmetic.multiply(value.re, value.im, scalar.re, scalar.im)))
)

/**
 * Lifts real values onto the complex plane.
 *
 * @since 0.1.0
 * @category conversions
 */
export const fromRealChunk = (values: Chunk.Chunk<number>): Chunk.Chunk<Complex> => Chunk.map(values, fromReal)

/**
 * Extracts real components from a complex vector.
 *
 * @since 0.1.0
 * @category conversions
 */
export const toRealChunk = (values: Chunk.Chunk<Complex>): Chunk.Chunk<number> => Chunk.map(values, (value) => value.re)

/**
 * Extracts imaginary components from a complex vector.
 *
 * @since 0.1.0
 * @category conversions
 */
export const toImaginaryChunk = (values: Chunk.Chunk<Complex>): Chunk.Chunk<number> =>
  Chunk.map(values, (value) => value.im)

/**
 * Computes each vector element's modulus.
 *
 * @since 0.1.0
 * @category conversions
 */
export const toMagnitudeChunk = (values: Chunk.Chunk<Complex>): Chunk.Chunk<number> => Chunk.map(values, abs)

/**
 * Computes each vector element's principal phase.
 *
 * @since 0.1.0
 * @category conversions
 */
export const toPhaseChunk = (values: Chunk.Chunk<Complex>): Chunk.Chunk<number> => Chunk.map(values, arg)

/**
 * Adds finite Cartesian operands after decoding with excess properties rejected.
 *
 * @since 0.1.0
 * @category validated operations
 */
export const addValidated = (input: unknown) =>
  Effect.map(
    decode(BinaryInput, "add", input),
    (value) => fromCartesian(Arithmetic.add(value.aRe, value.aIm, value.bRe, value.bIm))
  )

/**
 * Subtracts the second decoded finite operand from the first.
 *
 * @since 0.1.0
 * @category validated operations
 */
export const subtractValidated = (input: unknown) =>
  Effect.map(
    decode(BinaryInput, "subtract", input),
    (value) => fromCartesian(Arithmetic.subtract(value.aRe, value.aIm, value.bRe, value.bIm))
  )

/**
 * Multiplies finite Cartesian operands after strict boundary decoding.
 *
 * @since 0.1.0
 * @category validated operations
 */
export const multiplyValidated = (input: unknown) =>
  Effect.map(
    decode(BinaryInput, "multiply", input),
    (value) => fromCartesian(Arithmetic.multiply(value.aRe, value.aIm, value.bRe, value.bIm))
  )

/**
 * Divides decoded finite operands. A zero divisor still produces `NaN`
 * components, matching the trusted operation.
 *
 * @since 0.1.0
 * @category validated operations
 */
export const divideValidated = (input: unknown) =>
  Effect.map(
    decode(BinaryInput, "divide", input),
    (value) => fromCartesian(Arithmetic.divide(value.aRe, value.aIm, value.bRe, value.bIm))
  )

/**
 * Computes an exponential after decoding finite Cartesian components.
 *
 * @since 0.1.0
 * @category validated operations
 */
export const expValidated = (input: unknown) =>
  Effect.map(decode(Input, "exp", input), (value) => fromCartesian(Arithmetic.exp(value.re, value.im)))

/**
 * Computes the principal logarithm after decoding finite components. Zero is
 * accepted and maps to `-Infinity + 0i`.
 *
 * @since 0.1.0
 * @category validated operations
 */
export const logValidated = (input: unknown) =>
  Effect.map(decode(Input, "log", input), (value) => fromCartesian(Arithmetic.log(value.re, value.im)))

/**
 * Computes a modulus under runtime precision and diagnostic policies. Strict
 * precision rejects a non-finite result; diagnostics include input and result.
 *
 * @since 0.1.0
 * @category policy-aware operations
 */
export const absWithPolicies = (value: Complex) =>
  PolicyGuard.scalar({
    operation: "Complex.absWithPolicies",
    compute: () => abs(value),
    makeError: (message) => new DomainViolationError({ operation: "absWithPolicies", message }),
    annotations: (result) => ({
      input: Array.join(Array.make(encodeNumber(value.re), "+", encodeNumber(value.im), "i"), ""),
      result: encodeNumber(result)
    })
  })

/**
 * Computes a principal phase under runtime precision and diagnostic policies.
 * Strict precision rejects a non-finite result.
 *
 * @since 0.1.0
 * @category policy-aware operations
 */
export const argWithPolicies = (value: Complex) =>
  PolicyGuard.scalar({
    operation: "Complex.argWithPolicies",
    compute: () => arg(value),
    makeError: (message) => new DomainViolationError({ operation: "argWithPolicies", message }),
    annotations: (result) => ({
      input: Array.join(Array.make(encodeNumber(value.re), "+", encodeNumber(value.im), "i"), ""),
      result: encodeNumber(result)
    })
  })
